geode-core/src/test/java/org/apache/geode/cache/query/internal/QCompilerTest.java

```diff
@@ -0,0 +1,484 @@
+    assertThat(s3.compareTo(s2) > 0).isTrue();
+    assertThat(s4.compareTo(s3) > 0).isTrue();
+    assertThat(s5.compareTo(s4) > 0).isTrue();
+    assertThat(s6.compareTo(s5) > 0).isTrue();
+    assertThat(s7.compareTo(s6) > 0).isTrue();
+  }
+
+  @Test
+  @Parameters({"MIN", "MAX", "AVG", "SUM", "COUNT"})
+  public void parsingShouldSucceedWhenAggregatesAreUsedAsPartOfTheWhereClauseWithinAnInnerSelectQuery(
+      String function) {
+    String[] queries = new String[] {
+        "SELECT * FROM /region WHERE id IN (SELECT " + function + "(id) FROM /region)",
+        "SELECT COUNT(id), status FROM /region WHERE id IN (SELECT " + function
+            + "(id) FROM /region) GROUP BY status",
+
+        "SELECT * FROM /region WHERE id = ELEMENT(SELECT " + function + "(id) FROM /region)",
+        "SELECT * FROM /region WHERE id < ELEMENT(SELECT " + function + "(id) FROM /region)",
+        "SELECT * FROM /region WHERE id > ELEMENT(SELECT " + function + "(id) FROM /region)",
+        "SELECT * FROM /region WHERE id <> ELEMENT(SELECT " + function + "(id) FROM /region)",
+
+        "SELECT COUNT(id), status FROM /region WHERE id = ELEMENT(SELECT " + function
+            + "(id) FROM /region) GROUP BY status",
+        "SELECT COUNT(id), status FROM /region WHERE id < ELEMENT(SELECT " + function
+            + "(id) FROM /region) GROUP BY status",
+        "SELECT COUNT(id), status FROM /region WHERE id > ELEMENT(SELECT " + function
+            + "(id) FROM /region) GROUP BY status",
+        "SELECT COUNT(id), status FROM /region WHERE id <> ELEMENT(SELECT " + function
+            + "(id) FROM /region) GROUP BY status",
+    };
+
+    for (String queryString : queries) {
+      QCompiler compiler = new QCompiler();
+      assertThatCode(() -> compiler.compileQuery(queryString))
+          .as(String.format("Query parsing failed for %s but should have succeeded.", queryString))
+          .doesNotThrowAnyException();
+    }
+  }
+
+  @Test
+  @Parameters({"MIN", "MAX", "AVG", "SUM", "COUNT"})
+  public void parsingShouldThrowExceptionWhenAggregatesAreUsedAsPartOfTheWhereClauseWithoutAnInnerSelectQuery(
+      String function) {
+    String[] queries = new String[] {
+        "SELECT * FROM /region WHERE " + function + "(id) > 0",
+        "SELECT COUNT(id), status FROM /region WHERE " + function + "(id) > 0 GROUP BY status",
+        "SELECT * FROM /region WHERE " + function + "(id) > 0 OR id < 0",
+        "SELECT COUNT(id), status FROM /region WHERE " + function
+            + "(id) > 0 OR id < 0 GROUP BY status",
+        "SELECT * FROM /region WHERE id > 0 AND " + function + "(id) > 0",
+        "SELECT COUNT(id), status FROM /region WHERE id > 0 AND " + function
+            + "(id) > 0 GROUP BY status",
+        "SELECT * FROM /region WHERE " + function + "(id) < ELEMENT(SELECT " + function
+            + "(id) FROM /region)",
+        "SELECT COUNT(id), status FROM /region WHERE " + function + "(id) < ELEMENT(SELECT "
```

Comment:
one more test with outer query not having aggregate but inner query having aggregate in its where clause.

Reply:
Good idea, thanks.